### DISCUSSION TITLE
Make all operations leave a value on the stack for easier specialization.

### DISCUSSION THREAD

**Discussion by markshannon** [faster-cpython/ideas#241] [Ideas]
Most operation produce a value, and thus leave a value on the stack.
This has the nice property that the operation can be substituted with a call which has the same effect.
For example, we already specialize `BINARY_SUBSCR` with `BINARY_SUBSCR_GETITEM` for instances of classes that implement `__getitem__` in Python. `BINARY_SUBSCR_GETITEM` pushes a frame and then enters the `__getitem__` function without having to make any C calls.
We can do the same for `LOAD_ATTR`, and any calls (it makes sense to substitute one callable for another in some circumstances).

However, we cannot do easily do this for `__setitem__` or `__setattr__`. 
By changing `STORE_SUBSCR` and `STORE_ATTR` to leave a value on the stack, we can freely substitute them with a call.
Obviously changing `STORE_SUBSCR` to push `None` and then following it with `POP_TOP` is not very efficient if done naively.

However, it we guarantee that `STORE_SUBSCR` and `STORE_ATTR` are *always* followed by `POP_TOP`, we can simply change them to bump `next_instr` by one and skip the following instruction at no cost.

The question is: does the performance gain and simplification of the specialized forms outweigh the small cost of the extra 2 bytes per store?

I expect that it will, but we will need to experiment.

NOTE:
We can do the specialization without changing `STORE_SUBSCR`, but it requires the specialized form to push an extra frame just to execute an instruction to discard the top of the stack and pop the current frame: a `RETURN_VOID` instruction.

**Comment by gvanrossum** [comment]
Are you sure it's a perf gain?

**Comment by markshannon** [comment]
Am I sure? No:
 > I expect that it will, but we will need to experiment.
